Generalize ensemble datasets

aulemahal opened the pull request:
## Overview

This PR fixes #247, fixes #251.

Changes (breaking changes in bold):

* Rewrote all the management of datasets so that everything is parsed from a configuration file (`datasets.yml`).
* **Renamed the arguments to match the change: `rcp -> scenario`, `dataset_name -> dataset`.** (The latter was broken anyway).
* **Support for `candcs-u5` and `candcs-u6`**.  The name "bccaqv2" is translated into  `candcs-u5`.
* Because the CanDCS-U6 files have a deeper structure than the flat CanDCS-U5 list, the parsing methods have been generalized to recursively dig into subfolders (up to a given depth).
* The "allowed_values" of the WPS processes (args: `scenario, models`) are the union of all values of all datasets. However, the computation function will raise a "ValueError" if it receives arguments invalid for the chosen dataset. This means that currently, two kinds of errors are raised even though it's the same principle, I'll look into that.
* In theory, we could add any ensemble dataset even if they include more variables than `tasmax, tasmin, pr`. The code is currently generalized and the presence of new variables should trigger the wrapping of the additional compatible indicators,
* **Output filenames will be prefixed by the dataset name.** (unless a custom name is requested)
* The selection function accepts a list of string for `models`, but also a list of tuples (model name, realization). This is used for the PCIC12 "sublist", but we could make it visible to the user?
* There aren't global variable lists anymore.  The available variables are dataset-dependent and so are the compatible indicator. The ensemble common handler checks if the requested indicator is compatible with the request dataset. 
     - I removed `finch.processes.xclim` as it made no sense with this change.
* Removed `finch.processes.constants` because there was only 1 constant left.

## Related Issue / Discussion
@tlogan2000  Is there a sub-list for CanDCS-U6, similar to PCIC12?

## Additional Information
~I still need to update the doc, I'll add a section on how to pass dataset configuration.~

@perronld For local files, the YAML will look like:
```yaml
candcs-u5:
  local: true
  depth: 0
  path: /path/to/the/dataset/
  suffix: "*.nc"  # This is only used for local files
  pattern: "{variable}_{frequency}_BCCAQv2+ANUSPLIN300_{model}_{scenario}_{realization}_{date_start}-{date_end}.nc"
#  [the rest is the same]
```
The tests use a specific yaml that works with local files. See the documentation for details on the yaml.

aulemahal commented:
To illustrate the changes in the processes, here is the birdy-generated docstring of "ensemble_grid_point_cwd"
```
Returns the maximum number of consecutive wet days.

Parameters
----------
  [...]
dataset : {'bccaqv2', 'candcs-u5', 'candcs-u6'}string
    Name of the dataset from which to get netcdf files for inputs. 'BCCAQv2' redirects to CanDCS-U5 for backward compatibility.
scenario : {'rcp26', 'rcp45', 'rcp85', 'ssp126', 'ssp245', 'ssp585'}string
    Emission scenario (RCPs or SSPs, depending on the dataset)
models : {'BNU-ESM', 'CCSM4', 'CESM1-CAM5', 'CNRM-CM5', 'CSIRO-Mk3-6-0', 'CanESM2', 'FGOALS-g2', 'GFDL-CM3', 'GFDL-ESM2G', 'GFDL-ESM2M', ...}string
    When calculating the ensemble, include only these models. Allowed values depend on the dataset chosen. By default, all models are used, taking the first realization of each. Special sublists are also available :candcs-u5: dict_keys(['24models', 'pcic12']), candcs-u6: dict_keys(['26models'])
 [...]
**

aulemahal commented:
Well, my changes are more breaking than I first thought, especially when it comes to the tests.
I'm trying to make the available processes depend on the configurated datasets and which variable they provide,  but this choice has been quite costly... Anyway, I'm almost done

aulemahal commented:
This is now ready for review.

aulemahal commented:
![image](https://user-images.githubusercontent.com/20629530/192919138-f7ef8be6-ab6b-4a1a-aea9-6b96e5b20e97.png)
:face_exhaling: , :arrow_forward:  :sleeping_bed:

Zeitsperre commented:
@aulemahal A new CLISOPS will be available over the weekend. Sorry for the frustration!

Zeitsperre commented:
wat.
```
_________________ docs/source/notebooks/subset.ipynb::Cell 16 __________________
Notebook cell execution failed
Cell 16: Cell outputs differ

Input:
all_shapes.plot()

Traceback:
 mismatch 'text/plain'

 assert reference_output == test_output failed:

  '<AxesSubplot:>' == '<AxesSubplot: >'
  - <AxesSubplot: >
  ?              -
  + <AxesSubplot:>
```

tlogan2000 reviewed (approved):
I think this looks good.  I'm not sure how critical notebooks failures are on github actions

Zeitsperre commented:
I've opened a PR to modify the sanitization file (I ran into another issue concerning figure sizes when refreshing the notebooks on my work computer). I can add another rule for this `AxesSubplot` silliness to that PR as well and then we should be in the clear. Will modify and wait for @tlvu to review it before rerunning the checks here.

matprov commented on finch/processes/virtual.py:
```diff
@@ -0,0 +1,34 @@
+Processes as a virtual submodule
+--------------------------------
+When imported, this module exposes the default list of processes as virtual modules : indicators, ensemble and other.
+It forces the loading of the "default.cfg" file is no datasets were configured.
```
typo : "is" -> "if"

matprov commented on finch/processes/wps_xsubset_bbox_dataset.py:
```diff
@@ -69,11 +67,15 @@ def _handler(self, request: WPSRequest, response: ExecuteResponse):
+            dataset = get_datasets_config()[dataset_name]
         request.inputs["resource"] = get_datasets(
-            dataset_name, workdir=self.workdir, variables=variables, rcp=rcp
+            dataset, workdir=self.workdir,
```
@aulemahal I think that if `dataset_name != 'bccaqv2'`, `dataset` will not have been initialized. Is it possible that we need to put the `dataset = get_datasets_config()[dataset_name]` outside of the conditional?

huard replied:
I would have thought this should have caused a test failure, no?

aulemahal replied:
We don't test all processes with all datasets. And I probably didn't change this specific process' test (which was in itself a proof that my changes did work), so it might still be using `bccaqv2`...

I'll try to use `candcs-u5` everywhere in the tests.

aulemahal replied:
Woups, looks like the Dataset subset processes aren't tested anywhere! I'll add tests.

aulemahal replied:
@huard 
Évidemment, the dataset subset processes were broken. And not only because of this PR! I believe the CSV output was just not working on multi-point or bbox subsets. It is now fixed.

perronld replied:
If fixed, most likely this one can be close with the completion of this PR: https://github.com/bird-house/finch/issues/251

aulemahal replied:
Oh! Indeed, I forgot about that issue!

matprov commented on finch/processes/wps_xsubset_point_dataset.py:
```diff
@@ -94,11 +92,16 @@ def _handler(self, request: WPSRequest, response: ExecuteResponse):
-        dataset_name = single_input_or_none(request.inputs, "dataset_name")
+        dataset_name = single_input_or_none(request.inputs, "dataset")
+        if dataset_name == 'bccaqv2':
+            dataset_name = 'candcs-u5'
```
@aulemahal Do we plan to support this conversion in the long run, or is it temporary?

aulemahal replied:
I'm not sure...I'd say we keep it only temporarily, but I don't know who the users are really. If on your side, you never need the fix, I'll just remove in the next release.

The "problem" is that I can't make a "FutureWarning" or similar, so it will simply break for the users (I'm not sure there are many, though).

matprov replied:
Do you think that supporting the conversion will need other modification than the ones done in this PR?

huard replied:
Why not just make a copy of the yaml config for candcs-u5 called bccaqv2. You would not have to special case anything.

aulemahal replied:
yo.
Hacky config instead of hacky code? I'm in.

matprov commented on finch/processes/wps_xsubset_point_dataset.py:
```diff
@@ -94,11 +92,16 @@ def _handler(self, request: WPSRequest, response: ExecuteResponse):
         variable = request.inputs["variable"][0].data
         variables = None if variable is None else [variable]
-        rcp = single_input_or_none(request.inputs, "rcp")
+        scenario = single_input_or_none(request.inputs, "scenario")
```
@aulemahal Would it make sense to have backward compatibility on `rcp` and `dataset_name` (use `rcp` and `dataset_name` if `scenario` or `dataset` isn't defined)? This way this PR would not represent a breaking change and related to renamed fields.

aulemahal replied:
That's a bit heavier than the `bccaqv2` compatibility. It would mean having the two arguments in each processes? "scenario" + "rcp"?  I'll look into this.

Also, `dataset_name` was broken, it simply didn't work, at least for the indicator ensemble processes. So I'm guessing it isn't a breaking change?

matprov replied:
Forget about the compatibility if that's more work (and anyway doesn't make sense that `rcp` translates to `ssp` values). 
Changes related to `dataset_name`/`dataset` will be made on the portal itself so no worries about this change.

matprov commented:
@aulemahal Not sure why but I get a `PYWPS - ERROR - Exception: code: 400, description: Unknown process` for the following processes on the staging version using this branch.

```
subset_grid_point_dataset
subset_bbox_dataset
ensemble_grid_point_cold_spell_duration_index
ensemble_bbox_cold_spell_duration_index
ensemble_grid_point_heat_wave_frequency
```

matprov commented on CHANGES.rst:
```diff
@@ -1,6 +1,17 @@
+0.10.0 (unreleased)
+===================
+* Generalize ensemble datasets configuration.
+    - Datasets usable by ensemble processes are not specified through a YAML file which is pointed to in the configuration.
```
@aulemahal `are not specified`?

matprov commented on finch/processes/wps_xsubset_point_dataset.py:
```diff
@@ -94,11 +92,16 @@ def _handler(self, request: WPSRequest, response: ExecuteResponse):
+        dataset_name = single_input_or_none(request.inputs, "dataset")
+        if dataset_name == 'bccaqv2':
+            dataset_name = 'candcs-u5'
+        dataset = get_datasets_config()[dataset_name]
```
@aulemahal Failing to provide the "dataset" attribute in the request will result in a `KeyError: None`.

aulemahal replied:
Oh, I thought I had a default value! I'll make it so it defaults to the first dataset in the config (candcs-u5). (this will be documented in the process I believe)

aulemahal replied:
I instead chose to use the already existing `default_dataset` config value. Instead of being parsed in the process code, it is now used to fill the "default" field of the WPS input directly.

matprov reviewed (approved):
LGTM
Tested in CCDP's staging on finch.crim.ca with both CMIP5 and CMIP6.
Only failing test on my side is related to the comment on `wps_xsubset_bbox_dataset.py`.
Good job @aulemahal !

aulemahal commented on CHANGES.rst:
```diff
@@ -1,6 +1,17 @@
+0.10.0 (unreleased)
+===================
+* Generalize ensemble datasets configuration.
+    - Datasets usable by ensemble processes are not specified through a YAML file which is pointed to in the configuration.
```
`now` indeed.

Zeitsperre commented:
@aulemahal FYI the next xclim will be on conda-forge in the next few hours.

aulemahal commented:
@Zeitsperre  Nice! I'll do the update in another PR though, this new xclim has a large set of changes, I fear something might break.

aulemahal commented:
Ok! The last commits should fix the remaining bugs, but I did some non-trivial changes.

- There was a weird line in the netcdf to csv converter that would change the index of the first DataFrame to be concatenated. It seems like this line was breaking the conversion of any 3D (lat, lon, time) data. Which means it has been broken for a long time. It is now fixed, but I'm only 80% sure that all csv output will be the exact same as before.
- I removed `finch.wsgi.application`, I don't think it makes any sense to create an application on import because the user did not have the possibility to specify the configuration yet. This change can be reverted if needed.

matprov commented:
> I removed `finch.wsgi.application`, I don't think it makes any sense to create an application on import because the user did not have the possibility to specify the configuration yet. This change can be reverted if needed.

@aulemahal The subset box process now works, thanks. Please restore the `finch.wsgi.application` since otherwise it breaks the docker image. Thanks

aulemahal commented:
Ah merde, sorry. I'll do that asap!

tlvu commented on docs/source/notebooks/dap_subset.ipynb:
```diff
@@ -533,10 +534,10 @@
     "import datetime as dt\n",
     "coords = dict(lat=45, lon=290)\n",
-    "index, _ = xr.core.coordinates.remap_label_indexers(ds, coords, method=\"nearest\")\n",
+    "index = xr.core.indexing.map_index_queries(ds, coords, method=\"nearest\").dim_indexers\n",
```
@aulemahal Just FYI, Jenkins has been failing with error below since Nov 4?  Does it means the Jupyter env needs a newer version of xarray?  A new Jupyter env is hopefully coming soon so no big deal I guess.

If you change the tutorial notebooks in the future, please keep an eye on Jenkins or better, test with Jenkins in advance.  Jenkins can test against any devel branch.  I can show you how.

```
19:58:00  _________ finch-master/docs/source/notebooks/dap_subset.ipynb::Cell 2 __________
19:58:00  Notebook cell execution failed
19:58:00  Cell 2: Cell execution caused an exception
19:58:00  
19:58:00  Input:
19:58:00  # Use the `remap_label_indexers` function to convert coordinates to *positional* indexes.
19:58:00  import datetime as dt
19:58:00  coords = dict(lat=45, lon=290)
19:58:00  index = xr.core.indexing.map_index_queries(ds, coords, method="nearest").dim_indexers
19:58:00  
19:58:00  # The `nearest` method cannot be used with slices, so we do another call for the time period.
19:58:00  ti = xr.core.indexing.map_index_queries(ds, dict(time=slice("2060-01-01", "2064-12-30"))).dim_indexers
19:58:00  
19:58:00  # Merge the spatial and temporal indices
19:58:00  index.update(ti)
19:58:00  index
19:58:00  
19:58:00  Traceback:
19:58:00  
19:58:00  ---------------------------------------------------------------------------
19:58:00  AttributeError                            Traceback (most recent call last)
19:58:00  Input In [3], in <cell line: 4>()
19:58:00        2 import datetime as dt
19:58:00        3 coords = dict(lat=45, lon=290)
19:58:00  ----> 4 index = xr.core.indexing.map_index_queries(ds, coords, method="nearest").dim_indexers
19:58:00        6 # The `nearest` method cannot be used with slices, so we do another call for the time period.
19:58:00        7 ti = xr.core.indexing.map_index_queries(ds, dict(time=slice("2060-01-01", "2064-12-30"))).dim_indexers
19:58:00  
19:58:00  AttributeError: module 'xarray.core.indexing' has no attribute 'map_index_queries'
```

aulemahal replied:
Ach sorry about that.
Yes, I updated xarray in finch and this change is not backward compatible indeed.
